### ISSUE TITLE
ObjectDisposedException when function host shuts down

### ISSUE THREAD

**Issue by cgillum** [Azure/azure-functions-durable-extension#472]
Exception and call stack:

```
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'LoggerFactory'.
   at Microsoft.Extensions.Logging.LoggerFactory.CreateLogger(String categoryName)
   at Microsoft.Azure.WebJobs.Script.WebHost.ScriptLoggerFactory.CreateLogger(String categoryName) in C:\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Diagnostics\ScriptLoggerFactory.cs:line 26
   at Microsoft.Azure.WebJobs.Host.Loggers.FunctionInstanceLogger.Log(LogLevel level, FunctionDescriptor descriptor, Guid functionId, String message, Exception exception) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Loggers\FunctionInstanceLogger.cs:line 49
   at Microsoft.Azure.WebJobs.Host.Loggers.FunctionInstanceLogger.LogFunctionCompletedAsync(FunctionCompletedMessage message, CancellationToken cancellationToken) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Loggers\FunctionInstanceLogger.cs:line 36
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.TryExecuteAsync(IFunctionInstance functionInstance, CancellationToken cancellationToken) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:line 136
   at Microsoft.Azure.WebJobs.Host.Executors.TriggeredFunctionExecutor`1.TryExecuteAsync(TriggeredFunctionData input, CancellationToken cancellationToken) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\TriggeredFunctionExecutor.cs:line 54
   at Microsoft.Azure.WebJobs.Extensions.DurableTask.TaskActivityShim.RunAsync(TaskContext context, String rawInput)
```

Here is the kusto query where this issue was discovered:

```
FunctionsLogs 
| where PreciseTimeStamp between(datetime(2018-10-10 14:47) .. datetime(2018-10-10 14:50))
| where Tenant == "dc16b0dbc92d4dcca832b639e512cae6" and Pid == 10352
| where FunctionName == "" 
| project PreciseTimeStamp, Level, Summary, HostVersion, HostInstanceId
```
```
DurableFunctionsEvents 
| where PreciseTimeStamp between(datetime(2018-10-10 14:47) .. datetime(2018-10-10 14:50))
| where Tenant == "dc16b0dbc92d4dcca832b639e512cae6" and Pid == 10352
| where TaskName == "ExtensionInformationalEvent" or Level < 4
| where ProviderName == "WebJobs-Extensions-DurableTask"
| project TIMESTAMP, Pid, Level, TaskName, Details, Reason
```
The root cause is likely some kind of race condition when accessing the `LoggerFactory` object provided by the host.

**Comment by brettsam** [comment]
My guess is that this entire host is disposed (`LoggerFactory` tends to be the first thing that throws b/c it's used everywhere). Wasn't there an issue a while back where the Durable extension was caching the host instance (or some service) and re-using it?

**Comment by brettsam** [comment]
A hunch... note that I don't understand the full flow in durable :-)...

but when you call DurableTaskListener.StopAsync(), do we know that all of the Durable tasks have completed? I see that the calls to DeregisterOrchestrator/Activity set a property to true, but do those stop processing immediately, or is there a delay? If there's any delay there -- or if some functions have started and are not yet complete -- we can call dispose on all objects after exiting StopAsync(). My guess is that some of these are still in-flight and are being run against a disposed Executor.

**Comment by gorillapower** [comment]
Our Durable Function App was scaling out as we were doing a load test today and we encountered this exception.  

```
Unhandled exception while executing task: System.ObjectDisposedException: Cannot access a disposed object.\r\nObject name: 'LoggerFactory'.\r\n
   at Microsoft.Extensions.Logging.LoggerFactory.CreateLogger(String categoryName)\r\n
   at Microsoft.Azure.WebJobs.Script.WebHost.ScriptLoggerFactory.CreateLogger(String categoryName) in C:\\azure-webjobs-sdk-script\\src\\WebJobs.Script.WebHost\\Diagnostics\\ScriptLoggerFactory.cs:line 26\r\n
   at Microsoft.Azure.WebJobs.Host.Loggers.FunctionInstanceLogger.Log(LogLevel level, FunctionDescriptor descriptor, Guid functionId, String message, Exception exception) in C:\\projects\\azure-webjobs-sdk-rqm4t\\src\\Microsoft.Azure.WebJobs.Host\\Loggers\\FunctionInstanceLogger.cs:line 68\r\n
   at Microsoft.Azure.WebJobs.Host.Loggers.FunctionInstanceLogger.LogFunctionCompletedAsync(FunctionCompletedMessage message, CancellationToken cancellationToken) in C:\\projects\\azure-webjobs-sdk-rqm4t\\src\\Microsoft.Azure.WebJobs.Host\\Loggers\\FunctionInstanceLogger.cs:line 60\r\n
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.TryExecuteAsyncCore(IFunctionInstanceEx functionInstance, CancellationToken cancellationToken) in C:\\projects\\azure-webjobs-sdk-rqm4t\\src\\Microsoft.Azure.WebJobs.Host\\Executors\\FunctionExecutor.cs:line 158\r\n
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.TryExecuteAsync(IFunctionInstance instance, CancellationToken cancellationToken) in C:\\projects\\azure-webjobs-sdk-rqm4t\\src\\Microsoft.Azure.WebJobs.Host\\Executors\\FunctionExecutor.cs:line 87\r\n
   at Microsoft.Azure.WebJobs.Host.Executors.TriggeredFunctionExecutor`1.TryExecuteAsync(TriggeredFunctionData input, CancellationToken cancellationToken) in C:\\projects\\azure-webjobs-sdk-rqm4t\\src\\Microsoft.Azure.WebJobs.Host\\Executors\\TriggeredFunctionExecutor.cs:line 56\r\n
   at Microsoft.Azure.WebJobs.Extensions.DurableTask.TaskActivityShim.RunAsync(TaskContext context, String rawInput) in C:\\GitHub\\azure-functions-durable-extension\\src\\WebJobs.Extensions.DurableTask\\Listener\\TaskActivityShim.cs:line 57\r\n
   at DurableTask.Core.TaskActivityDispatcher.<>c__DisplayClass12_1.<<OnProcessWorkItemAsync>b__1>d.MoveNext() in C:\\source\\durabletask\\src\\DurableTask.Core\\TaskActivityDispatcher.cs:line 135\n\t
   at Microsoft.Extensions.Logging.LoggerFactory.CreateLogger(String categoryName)\r\n
   at Microsoft.Azure.WebJobs.Script.WebHost.ScriptLoggerFactory.CreateLogger(String categoryName) in C:\\azure-webjobs-sdk-script\\src\\WebJobs.Script.WebHost\\Diagnostics\\ScriptLoggerFactory.cs:line 26\r\n
   at Microsoft.Azure.WebJobs.Host.Loggers.FunctionInstanceLogger.Log(LogLevel level, FunctionDescriptor descriptor, Guid functionId, String message, Exception exception) in C:\\projects\\azure-webjobs-sdk-rqm4t\\src\\Microsoft.Azure.WebJobs.Host\\Loggers\\FunctionInstanceLogger.cs:line 68\r\n
   at Microsoft.Azure.WebJobs.Host.Loggers.FunctionInstanceLogger.LogFunctionCompletedAsync(FunctionCompletedMessage message, CancellationToken cancellationToken) in C:\\projects\\azure-webjobs-sdk-rqm4t\\src\\Microsoft.Azure.WebJobs.Host\\Loggers\\FunctionInstanceLogger.cs:line 60\r\n
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.TryExecuteAsyncCore(IFunctionInstanceEx functionInstance, CancellationToken cancellationToken) in C:\\projects\\azure-webjobs-sdk-rqm4t\\src\\Microsoft.Azure.WebJobs.Host\\Executors\\FunctionExecutor.cs:line 158\r\n
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.TryExecuteAsync(IFunctionInstance instance, CancellationToken cancellationToken) in C:\\projects\\azure-webjobs-sdk-rqm4t\\src\\Microsoft.Azure.WebJobs.Host\\Executors\\FunctionExecutor.cs:line 87\r\n
   at Microsoft.Azure.WebJobs.Host.Executors.TriggeredFunctionExecutor`1.TryExecuteAsync(TriggeredFunctionData input, CancellationToken cancellationToken) in C:\\projects\\azure-webjobs-sdk-rqm4t\\src\\Microsoft.Azure.WebJobs.Host\\Executors\\TriggeredFunctionExecutor.cs:line 56\r\n
   at Microsoft.Azure.WebJobs.Extensions.DurableTask.TaskActivityShim.RunAsync(TaskContext context, String rawInput) in C:\\GitHub\\azure-functions-durable-extension\\src\\WebJobs.Extensions.DurableTask\\Listener\\TaskActivityShim.cs:line 57\r\n
   at DurableTask.Core.TaskActivityDispatcher.<>c__DisplayClass12_1.<<OnProcessWorkItemAsync>b__1>d.MoveNext() in C:\\source\\durabletask\\src\\DurableTask.Core\\TaskActivityDispatcher.cs:line 135
```

**Comment by cgillum** [comment]
I spent quite some time investigating another instance of this issue and can confirm that this is a race condition when the functions host starts shutting down. It's especially problematic in the Azure Functions Consumption plan when there is a lot of scaling out and back in again.

To summarize, this issue is caused by the host shutting down while an activity function is still executing. If the activity function completes after the host has been stopped but before the process is recycled, the completed activity function can trigger this exception.

I'm going to prioritize fixing this issue since it's a reliability bug. It will be a little complicated to fix since it involves coordinating several moving pieces, but I expect to have a fix ready as part of v2.1.1 sometime within the month of January.
